### DISCUSSION TITLE
为什么 TCP 建立连接需要三次握手 · Why's THE Design?(003) · /whys-the-design-tcp-three-way-handshake

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#572] [Announcements]
https://draveness.me/whys-the-design-tcp-three-way-handshake

**Comment by ttttmr** [comment]
为什么DNS使用UDP？
也是面试问到了，我觉得是节省服务端资源，响应快，面试官好像觉得回答的不好，一直问还有呢，来请教一下博主

**Reply by xavierskip** [reply]
tcp在传输数据前要完成三次握手，也就是客户端要收到域名解析信息需要经过5次通信，udp的话，正常通信只要2次就可以收到了。

**Reply by xavierskip** [reply]
也就是 tcp 要两个RTT（往返时延），udp只要一个RTT就行。

**Reply by tico-zl** [reply]
简单和低延迟，不需要tcp复杂的握手；
效率高占用资源少，udp头部8个字节，tcp20个；
支持多播和广播；

**Comment by draveness** [comment]
> 为什么DNS使用UDP？
> 也是面试问到了，我觉得是节省服务端资源，响应快，面试官好像觉得回答的不好，一直问还有呢，来请教一下博主

这是一个非常好的问题，我会分析一下的，多谢

**Comment by beihai0xff** [comment]
为什么博客突然间无法访问了，能上谷歌，看着看着就无连接了

**Comment by draveness** [comment]
> 为什么博客突然间无法访问了，能上谷歌，看着看着就无连接了

现在好了么，可能是服务器网络抖动

**Comment by beihai0xff** [comment]
> > 为什么博客突然间无法访问了，能上谷歌，看着看着就无连接了
> 
> 现在好了么，可能是服务器网络抖动

现在可以了，这几天一直是这样，还以为是我梯子问题

**Comment by XiaTianliang** [comment]
> @ttttmr
> 为什么DNS使用UDP？
> 也是面试问到了，我觉得是节省服务端资源，响应快，面试官好像觉得回答的不好，一直问还有呢，来请教一下博主

dns交换的信息很短，为了这么简短的请求建立tcp觉得太浪费了。

**Comment by leafduo** [comment]
> @ttttmr
> 为什么DNS使用UDP？
> 也是面试问到了，我觉得是节省服务端资源，响应快，面试官好像觉得回答的不好，一直问还有呢，来请教一下博主

当时可能是节省时延和服务器资源之类的吧，现在来看 DNS 使用 UDP 的弊端暴露地越来越多了，比如安全、隐私和流量放大之类的问题

**Comment by leafduo** [comment]
看完就想到了 TCP Fast Open，好奇它是怎么不用三次握手然后做到拒绝重复连接的，结果它直接接受了重复连接，要求应用层自己判断只有幂等操作才用 Fast Open。这个复杂性可能也是推不下去的原因之一吧。

**Comment by geckofu** [comment]
DNS 用的是 Anycast IP 地址, 每次查询都可能和不同的机器通信, 所以没有办法维持状态

**Comment by ttttmr** [comment]
> DNS 用的是 Anycast IP 地址, 每次查询都可能和不同的机器通信, 所以没有办法维持状态

好像有点道理，那么问题又来了😂，DoH肯定是基于TCP，比如1.1.1.1的那个，这是怎么做到的呢

**Comment by houbaron** [comment]
请问这些图都是用什么软件绘制的？似乎看过类似的用 Sketch 画的，如果是 Sketch，能分享下素材吗

**Comment by geckofu** [comment]
> @ttttmr
> > DNS 用的是 Anycast IP 地址, 每次查询都可能和不同的机器通信, 所以没有办法维持状态
> 
> 好像有点道理，那么问题又来了😂，DoH肯定是基于TCP，比如1.1.1.1的那个，这是怎么做到的呢

我也不了解了 😂 还想到一个问题, 为什么 IP 可以上 HTTPS？不是域名才可以的么?

**Comment by leafduo** [comment]
> @geckofu
> DNS 用的是 Anycast IP 地址, 每次查询都可能和不同的机器通信, 所以没有办法维持状态

只有少数公共 DNS 是 anycast，并且实践中大部分连接是能保持连接状态的。

https://serverfault.com/questions/495719/anycast-dns-how-do-you-deal-with-tcp-dns-requests

**Comment by Maecenas** [comment]
> @ttttmr
> 为什么DNS使用UDP？
> 也是面试问到了，我觉得是节省服务端资源，响应快，面试官好像觉得回答的不好，一直问还有呢，来请教一下博主

DNS是一个比较简单的机制，这可能不是选择UDP的主要原因。TCP提供的、两个端点之间的可靠连接不是DNS需要的，DNS协议的可靠连接机制是靠多组冗余的DNS服务器实现的。DNS只有在需要可靠传输数据（zone transfer）和数据量较大（> 512 bytes）才会使用TCP。

**Comment by rsj217** [comment]
> 这种用类比来解释问题往往就会面临『十个类比九个错』的尴尬局面

是的，最开始学习写笔记的时候也喜欢这样类比，看起来貌似很`巧妙`，实则很多非但不能解释清楚问题，还带来新的困惑。

**Comment by draveness** [comment]
为什么 DNS 使用 UDP 协议 https://draveness.me/whys-the-design-dns-udp-tcp

cc/ @ttttmr

**Comment by yanjinbin** [comment]
哈哈 厉害了 最后2个提问 属于好问题 
1  tcp C端和S端 如何保证消息不丢失不重复，像FIN ACK这样的单调递增序列，还有吗？ 不知道
2 UDP 面向数据报，数据传输无序，也米有TCP拥塞控制，但是 我记得有丢失重发的机制 ，待会我找找

**Comment by geckofu** [comment]
<img width="1147" alt="Screen Shot 2019-11-14 at 8 43 43 PM" src="https://user-images.githubusercontent.com/4544326/68858255-97c6f300-071f-11ea-9642-ad0f79a64d5d.png">
博客证书是 Fake Certificate 啦 🤣 

---
Update:
很快又恢复了

**Comment by draveness** [comment]
刚才出了点问题...

**Comment by Utachen** [comment]
请教一下博主，用什么软件画的图，非常心水

**Comment by wusphinx** [comment]
很多类比忽略掉了细节，反而容易偏离问题的核心，笔者从**TCP连接**的定义为切入点，思路清奇，娓娓道来，让人受益匪浅

**Comment by bspp1314** [comment]
“类比带来的解释往往只能有片面的相似性，我们永远也无法找到绝对正确的类比，它只在我们想要通俗易懂地展示事物的特性时才能发挥较大的作用。”非常好的说法，当我们想正确且深刻的认识某一问题的时候，使用类比确实是一种偷懒的行为，只会给你以后解决这类问题带来更大的麻烦。

**Comment by wanglongbiao** [comment]
图片画的很漂亮

**Comment by abing0610** [comment]
为什么四次挥手

**Comment by yangsoon** [comment]
> 所以，TCP 选择使用三次握手来建立连接并在连接引入了 RST 这一控制消息，接收方当收到请求时会将发送方发来的 SEQ+1 发送回接收方 

**发送回接收方** 应该发送回发送方？或者说发送回对方

**Comment by draveness** [comment]
> > 所以，TCP 选择使用三次握手来建立连接并在连接引入了 RST 这一控制消息，接收方当收到请求时会将发送方发来的 SEQ+1 发送回接收方
> 
> **发送回接收方** 应该发送回发送方？或者说发送回对方

发送回对方比较准确

**Comment by yangsoon** [comment]
> @draveness
> > > 所以，TCP 选择使用三次握手来建立连接并在连接引入了 RST 这一控制消息，接收方当收到请求时会将发送方发来的 SEQ+1 发送回接收方
> > 
> > **发送回接收方** 应该发送回发送方？或者说发送回对方
> 
> 发送回对方比较准确

嗯嗯 希望博主能出一个讲解关于TCP三次握手能否传输数据的文章

**Comment by zxfendoubuzhi** [comment]
谢谢你, 这个系列文章太棒了.

**Comment by vanvanj** [comment]
用UDP如何实现可靠数据传输
面试经常被问, 求大神也分析一下

**Comment by draveness** [comment]
> 用UDP如何实现可靠数据传输
> 面试经常被问, 求大神也分析一下

可靠传输一般包含两个方面，数据不会丢失，数据保证顺序；

1. 数据不丢失依靠发送方的重传策略；
2. 数据保证顺序需要通信双方维护序列号；

遵循以上的两个策略就能实现一个最基本的可靠传输协议。

**Comment by ghost** [comment]
“当前连接是历史连接，即 SEQ 过期或者超时”关于这句话有两个问题：
1）请问kernel （server端) 是如何判断seq超时的，因为tcp header上并没有timeout或者MSL的设置；
2）如果kernel收到同一个client的两个seq，其ISN不一样，其处理逻辑是根据记录的client addr丢掉第二个接收的ISN么，不是重复建立第二个连接么？这个理论有没有文档支撑呢？

**Comment by ghost** [comment]
“这种增加 TCP 连接通信次数的问题往往没有讨论的必要性，我们追求的其实是用更少的通信次数（理论上的边界）完成信息的交换”

这个其实有一定的必要性，理论如下：
1）tcp的现实运转中确实存在4次的情况，考虑simultaneous open的情况，当然是小众；
2）对比一下sctp和带cookie的tcp模式下的握手次数，4次和3次有时候边界情况是模糊的；

**Comment by draveness** [comment]
> “当前连接是历史连接，即 SEQ 过期或者超时”关于这句话有两个问题：
> 1）请问kernel （server端) 是如何判断seq超时的，因为tcp header上并没有timeout或者MSL的设置；
> 2）如果kernel收到同一个client的两个seq，其ISN不一样，其处理逻辑是根据记录的client addr丢掉第二个接收的ISN么，不是重复建立第二个连接么？这个理论有没有文档支撑呢？

看下 RFC793 的 3.4.  Establishing a connection，最终判断 SEQ 是否过期的是发起方，这应该能解决这两个问题

> an old duplicate SYN arrives at TCP B.  TCP B
  cannot tell that this is an old duplicate, so it responds normally
  (line 4).  TCP A detects that the ACK field is incorrect and returns a
  RST (reset) with its SEQ field selected to make the segment
  believable.

**Comment by draveness** [comment]
> “这种增加 TCP 连接通信次数的问题往往没有讨论的必要性，我们追求的其实是用更少的通信次数（理论上的边界）完成信息的交换”
> 
> 这个其实有一定的必要性，理论如下：
> 1）tcp的现实运转中确实存在4次的情况，考虑simultaneous open的情况，当然是小众；
> 2）对比一下sctp和带cookie的tcp模式下的握手次数，4次和3次有时候边界情况是模糊的；

感谢你的认真，提的两个理论没错，但是跟这里的必要性是两件事。文中已经用了关键字**理论上的边界**，3 次是 TCP 双方无上下文时能够达成共识的最小次数，更多的次数 4 次、5 次和 6 次 都可以完成通信，这已经构成了 TCP 不需要 3 次以上通信的充分条件，所以没有讨论的必要性。

**Comment by ghost** [comment]
> > “这种增加 TCP 连接通信次数的问题往往没有讨论的必要性，我们追求的其实是用更少的通信次数（理论上的边界）完成信息的交换”
> > 这个其实有一定的必要性，理论如下：
> > 1）tcp的现实运转中确实存在4次的情况，考虑simultaneous open的情况，当然是小众；
> > 2）对比一下sctp和带cookie的tcp模式下的握手次数，4次和3次有时候边界情况是模糊的；
> 
> 感谢你的认真，提的两个理论没错，但是跟这里的必要性是两件事。文中已经用了关键字**理论上的边界**，3 次是 TCP 双方无上下文时能够达成共识的最小次数，更多的次数 4 次、5 次和 6 次 都可以完成通信，这已经构成了 TCP 不需要 3 次以上通信的充分条件，所以没有讨论的必要性。

不是哦，当双方不是passive-active模式下，并无法根据3次达成最小共识；3次只是在这个模式下讨论的最小次数，只是因为tcp常用于c-s模式所以主要讲3次不代表tcp不需要三次；

而且题主可以稍微思考一下我说的第二个点，sctp和带cookie的tcp设计可能会对多次有新的认识

**Comment by ghost** [comment]
> > “当前连接是历史连接，即 SEQ 过期或者超时”关于这句话有两个问题：
> > 1）请问kernel （server端) 是如何判断seq超时的，因为tcp header上并没有timeout或者MSL的设置；
> > 2）如果kernel收到同一个client的两个seq，其ISN不一样，其处理逻辑是根据记录的client addr丢掉第二个接收的ISN么，不是重复建立第二个连接么？这个理论有没有文档支撑呢？
> 
> 看下 RFC793 的 3.4. Establishing a connection，最终判断 SEQ 是否过期的是发起方，这应该能解决这两个问题
> 
> > an old duplicate SYN arrives at TCP B.  TCP B
> > cannot tell that this is an old duplicate, so it responds normally
> > (line 4).  TCP A detects that the ACK field is incorrect and returns a
> > RST (reset) with its SEQ field selected to make the segment
> > believable.

这个地方有点疑问，我刚刚review了一下RFC然后按照MSL由sender判断的话有两个疑问
这种sender判断的机制下，是不是一个client多次sync会占用两个connection的资源也就是server会在sync queue里enqueue两个incomplete 的connection；如果这样那么一个mallicious sender的DDoS攻击成本会很低（当然可以新增其他方案再改进这种攻击）

不过RFC对这个问题讲的比较笼统，而且几个例子要么在silmutanous open讲的要么在crash角度讲的，所以这里我的疑问比较大。

**Comment by draveness** [comment]
> 而且题主可以稍微思考一下我说的第二个点，sctp和带cookie的tcp设计可能会对多次有新的认识

3 次是 TCP 双方**无上下文时**能够达成共识的最小次数，带 Cookie 的就有是状态的。

**Comment by draveness** [comment]
> > > “当前连接是历史连接，即 SEQ 过期或者超时”关于这句话有两个问题：
> > > 1）请问kernel （server端) 是如何判断seq超时的，因为tcp header上并没有timeout或者MSL的设置；
> > > 2）如果kernel收到同一个client的两个seq，其ISN不一样，其处理逻辑是根据记录的client addr丢掉第二个接收的ISN么，不是重复建立第二个连接么？这个理论有没有文档支撑呢？
> > 
> > 
> > 看下 RFC793 的 3.4. Establishing a connection，最终判断 SEQ 是否过期的是发起方，这应该能解决这两个问题
> > > an old duplicate SYN arrives at TCP B.  TCP B
> > > cannot tell that this is an old duplicate, so it responds normally
> > > (line 4).  TCP A detects that the ACK field is incorrect and returns a
> > > RST (reset) with its SEQ field selected to make the segment
> > > believable.
> 
> 这个地方有点疑问，我刚刚review了一下RFC然后按照MSL由sender判断的话有两个疑问
> 这种sender判断的机制下，是不是一个client多次sync会占用两个connection的资源也就是server会在sync queue里enqueue两个incomplete 的connection；如果这样那么一个mallicious sender的DDoS攻击成本会很低（当然可以新增其他方案再改进这种攻击）

按照 RFC793 的介绍确实可能会有这种问题，但是这不在这篇文章的范围内。

**Comment by ghost** [comment]
嗯，是基于三次握手延伸思考没有很确定答案所以才问的，只局限于文章内部提到的信息就比较简单

**Comment by draveness** [comment]
> 嗯，是基于三次握手延伸思考没有很确定答案所以才问的，只局限于文章内部提到的信息就比较简单

👍 手动点赞

**Comment by ghost** [comment]
"最终判断 SEQ 是否过期的是发起方”

这个的点其实想讨论一下，因为看了一些教材没想明白。如果是这个逻辑的话，time-wait状态感觉并无法防止delayed segments被新的incarnation connection（镜像连接）混淆。

推理过程是：1）在新的连接seq number也符合window，这个时候网络上的segment被对方回传了（基于是否过期是发起方来判断这个机制，对方收到的超时很久的segment也得回复靠发起方判断）；
2）然而这个时候发起方和对方已经是一个新的connection了，不再是time-wait状态了；3）由于这个segment上没有timestamp也没有msl这些信息，发起方怎么判断是否是上次的包呢，
2MSL的time-wait这样看来从时间上对解决连接混淆的问题看起来好像没什么作用？

**Comment by draveness** [comment]
> "最终判断 SEQ 是否过期的是发起方” 这个的点其实想讨论一下，因为看了一些教材没想明白。如果是这个逻辑的话，time-wait状态感觉并无法防止delayed segments被新的incarnation connection（镜像连接）混淆。 推理过程是：1）在新的连接seq number也符合window，这个时候网络上的segment被对方回传了（基于是否过期是发起方来判断这个机制，对方收到的超时很久的segment也得回复靠发起方判断）； 2）然而这个时候发起方和对方已经是一个新的connection了，不再是time-wait状态了；3）由于这个segment上没有timestamp也没有msl这些信息，发起方怎么判断是否是上次的包呢， 2MSL的time-wait这样看来从时间上对解决连接混淆的问题看起来好像没什么作用？

不好意思回复晚了，不过我看了很多遍都没看懂你的问题😂😂😂

你说的问题是在这个 [RFC1337](https://tools.ietf.org/html/rfc1337) 里的么

**Comment by renyuan431** [comment]
请博主收下我的膝盖，看完感觉醍醐灌顶，茅塞顿开，但有个概念不知道理解的是否正确，麻烦博主在百忙之中帮忙指点一二。
“除此之外，网络作为一个分布式的系统，其中并不存在一个用于计数的全局时钟，而 TCP 可以通过不同的机制来初始化序列号，作为 TCP 连接的接收方我们无法判断对方传来的初始化序列号是否过期，所以我们需要交由对方来判断，TCP 连接的发起方可以通过保存发出的序列号判断连接是否过期”这段话是说，连接的双方都无法知道对方是什么时候发送的TCP握手的通信，因为两边没有一个统一的时钟，你的8点和我的8点完全不是一个时间，所以只有自己这端知道自己的信息是什么时候发送的，是否为真正那个建立连接的信息，所以自己发出的信息是否过期只有自己才知道才能判断，不知道我理解的是否正确。

**Comment by draveness** [comment]
> “除此之外，网络作为一个分布式的系统，其中并不存在一个用于计数的全局时钟，而 TCP 可以通过不同的机制来初始化序列号，作为 TCP 连接的接收方我们无法判断对方传来的初始化序列号是否过期，所以我们需要交由对方来判断，TCP 连接的发起方可以通过保存发出的序列号判断连接是否过期”这段话是说，连接的双方都无法知道对方是什么时候发送的TCP握手的通信，因为两边没有一个统一的时钟，你的8点和我的8点完全不是一个时间，所以只有自己这端知道自己的信息是什么时候发送的，是否为真正那个建立连接的信息，所以自己发出的信息是否过期只有自己才知道才能判断，不知道我理解的是否正确。

时钟和时间不太一样，这里说的其实就是唯一的序列号，只有发送方才能存储并验证自己发出的序列号是不是过期的，除了这里，其他的理解应该没啥问题。

**Comment by LuckyDevin** [comment]
不一样的视角，很清晰

**Comment by sijunzi** [comment]
可以告诉我你的博客是用什么做的么？

**Comment by draveness** [comment]
> 可以告诉我你的博客是用什么做的么？

Hugo

**Comment by zhxilin** [comment]
> @draveness
> > “当前连接是历史连接，即 SEQ 过期或者超时”关于这句话有两个问题：
> > 1）请问kernel （server端) 是如何判断seq超时的，因为tcp header上并没有timeout或者MSL的设置；
> > 2）如果kernel收到同一个client的两个seq，其ISN不一样，其处理逻辑是根据记录的client addr丢掉第二个接收的ISN么，不是重复建立第二个连接么？这个理论有没有文档支撑呢？
> 
> 看下 RFC793 的 3.4.  Establishing a connection，最终判断 SEQ 是否过期的是发起方，这应该能解决这两个问题
> 
> > an old duplicate SYN arrives at TCP B.  TCP B
>   cannot tell that this is an old duplicate, so it responds normally
>   (line 4).  TCP A detects that the ACK field is incorrect and returns a
>   RST (reset) with its SEQ field selected to make the segment
>   believable.
> 
> 

借博主的图补充解释一下，TCP A假设发送了最新Seq=100的包，但是此前Seq=90的包更早被TCP B接收，然后回复TCP A AckNum=90+1的包，TCP A收到之后，比较上下文发现自己期望的回复AckNum应为100+1而不是90+1，认定为这是一个历史连接，所以会发起RST终止旧的连接

**Comment by liujiejesse** [comment]
点赞支持

**Comment by TigerZhag** [comment]
您好，如果B发送了(SYN,ACK)数据包后, 长时间没有得到A的回复, 是否会重新发送呢, 如果不重新发送的话，什么时候放弃此连接。

**Comment by draveness** [comment]
> 您好，如果B发送了(SYN,ACK)数据包后, 长时间没有得到A的回复, 是否会重新发送呢, 如果不重新发送的话，什么时候放弃此连接。

会有超时和重试次数的

**Comment by TigerZhag** [comment]
> @draveness
> > 您好，如果B发送了(SYN,ACK)数据包后, 长时间没有得到A的回复, 是否会重新发送呢, 如果不重新发送的话，什么时候放弃此连接。
> 
> 会有超时和重试次数的

如果重试的话，A收到了历史SYN,ACK数据包，又发送了ACK给B，这时B判断此连接已过期，会发送RST，然后可能A收到了最新的SYN，ACK数据包，再次给B发送ACK，连接就建立了，这样理解对吗

**Comment by draveness** [comment]
> > @draveness
> > > 您好，如果B发送了(SYN,ACK)数据包后, 长时间没有得到A的回复, 是否会重新发送呢, 如果不重新发送的话，什么时候放弃此连接。
> > 
> > 
> > 会有超时和重试次数的
> 
> 如果重试的话，A收到了历史SYN,ACK数据包，又发送了ACK给B，这时B判断此连接已过期，会发送RST，然后可能A收到了最新的SYN，ACK数据包，再次给B发送ACK，连接就建立了，这样理解对吗

我其实没有完全看懂你的描述

**Comment by TigerZhag** [comment]
> @draveness
> > > @draveness
> > > > 您好，如果B发送了(SYN,ACK)数据包后, 长时间没有得到A的回复, 是否会重新发送呢, 如果不重新发送的话，什么时候放弃此连接。
> > > 
> > > 
> > > 会有超时和重试次数的
> > 
> > 如果重试的话，A收到了历史SYN,ACK数据包，又发送了ACK给B，这时B判断此连接已过期，会发送RST，然后可能A收到了最新的SYN，ACK数据包，再次给B发送ACK，连接就建立了，这样理解对吗
> 
> 我其实没有完全看懂你的描述

就是文中只提到了B回复(CTL=SYN,ACK)时，如果A发现SEQ已过期，会回复RST，并重新连接，我确认一下在第三次握手时，如果B发现A的SEQ已过期，是不是同样的处理方式。

产生疑问的原因是，A才是连接发起方，如果B发送RST的话，那重新连接是谁来发起呢
不知道这次有没有表达清楚，感谢解惑

**Comment by TigerZhag** [comment]
还有一个疑问，在介绍历史错误连接的配图里，为什么是第二次发送的SEQ=90的那个请求duplicate了，而不是最早发送的SEQ=100的请求，是不是画错了

**Comment by oldthreefeng** [comment]
@draveness 文章转载了~ 写的特别好。

**Comment by draveness** [comment]
> @draveness 文章转载了~ 写的特别好。

文章下面有如何转载的提示

**Comment by 879913935** [comment]
除了使用序列号是否还有其他方式保证消息的不重不丢？

这个没理解，tcp如果没有序号，还能用什么方式实现

**Comment by zinwalin** [comment]
{% include related/whys-the-design.md %}

**Comment by Gpia** [comment]
从标准定义看，连接包含：Socket、序列号、窗口大小和一些控制信息。
发送方和接收方协商的只有序列号和控制信息。这里的socket和窗口大小的初始化是发送方自己定的吧。接收方一直是listen状态。

**Comment by hzh-test** [comment]
发送端发送连接请求如果没有收到确认，重发序列号是相同的（使用curl请求，wireshark抓包已证明），但是重试次数是有限的，之后用户可能会手动再次触发新的连接请求，此时的序列号是不同的，这时接受端就有可能收到多个不同的连接请求，序列号相同接受端可以辨认，但是这里是存在两个不同序列号的请求。如果只有两次握手，接收端就需要维护两个连接，但其中有一个是无用的，这就浪费资源了。这和 DDOS 攻击最大的区别是这个是非人为的无意的，但却会造成更大的危害。另外，发送端是可以伪造源IP地址的，这种接收端连的其实都不是发送端，这还能叫连接吗。
究其根源，是网络可不可达导致的问题。从发送端发出连接请求到发送端收到消息确认，在客户端看来确实是证明了网络可达并且连接已经建立，但是在接收端看来，单单收到连接请求并不能证明什么，因为这个连接请求有可能是旧的延迟的连接请求，此时客户端可能早就下线了。所以从接收端的角度来看，其也需要确认网络可达（或者说接收端在线），所以客户端收到消息确认后还得再回复一下接收端，这样一来一回就实现了双向证明。
综上，就算发送端伪造源IP地址，连接也无法建立，因为接收端发送的消息客户端收不到，自然也就不会再收到来自客户端的确认消息。另外，非人为的无意的类似 DDOS 攻击的行为也不会发生了，两全其美，一箭双雕。

这个问题面试遇到过，当时也只是在回答三次握手的最后一步，之后越想越不对劲，搜了很多答案都不满意，今天看到博主的文章，终于是彻底地把自己给说服了。

**Comment by nonacosa** [comment]
大佬用的这是什么画图工具？

**Comment by Dcell** [comment]
希望大佬能做一篇 http 协议的进化史，非常期待

**Comment by xinnjie** [comment]
> @hzh-test
> 发送端发送连接请求如果没有收到确认，重发序列号是相同的（使用curl请求，wireshark抓包已证明），但是重试次数是有限的，之后用户可能会手动再次触发新的连接请求，此时的序列号是不同的，这时接受端就有可能收到多个不同的连接请求，序列号相同接受端可以辨认，但是这里是存在两个不同序列号的请求。如果只有两次握手，接收端就需要维护两个连接，但其中有一个是无用的，这就浪费资源了。这和 DDOS 攻击最大的区别是这个是非人为的无意的，但却会造成更大的危害。另外，发送端是可以伪造源IP地址的，这种接收端连的其实都不是发送端，这还能叫连接吗。
> 究其根源，是网络可不可达导致的问题。从发送端发出连接请求到发送端收到消息确认，在客户端看来确实是证明了网络可达并且连接已经建立，但是在接收端看来，单单收到连接请求并不能证明什么，因为这个连接请求有可能是旧的延迟的连接请求，此时客户端可能早就下线了。所以从接收端的角度来看，其也需要确认网络可达（或者说接收端在线），所以客户端收到消息确认后还得再回复一下接收端，这样一来一回就实现了双向证明。
> 综上，就算发送端伪造源IP地址，连接也无法建立，因为接收端发送的消息客户端收不到，自然也就不会再收到来自客户端的确认消息。另外，非人为的无意的类似 DDOS 攻击的行为也不会发生了，两全其美，一箭双雕。
> 
> 这个问题面试遇到过，当时也只是在回答三次握手的最后一步，之后越想越不对劲，搜了很多答案都不满意，今天看到博主的文章，终于是彻底地把自己给说服了。
> 

手动点赞

**Comment by Maecenas** [comment]
> @birdkyle7918
> 作者你好，请问你的图是用什么软件画的呀？很漂亮

最后一行：文章图片
你可以在 [技术文章配图指南](https://draveness.me/sketch-and-sketch/) 中找到画图的方法和素材。

**Comment by uknfire** [comment]
总结来讲，在TCP握手中，因为发起方知道更多的背景信息，所以应该由发起方主导控制这个连接。
因此协议给了发起方第二次发送消息的机会，根据已知的信息发送ACK接受连接，或者发送RST中止连接

**Comment by xincheng1997** [comment]
除了使用序列号是否还有其他方式保证消息的不重不丢？
有哪些实现呢？

**Comment by BBplux** [comment]
三次握手有了怎么能少了四次挥手

**Comment by vc60er** [comment]
期待博主的四次挥手
